Packages/com.unity.inputsystem/InputSystem/Actions/InputActionMap.cs

```diff
@@ -739,13 +693,84 @@ IEnumerator IEnumerable.GetEnumerator()
         [NonSerialized] private bool m_NeedToResolveBindings;
         [NonSerialized] internal InputBinding? m_BindingMask;
 
-        [NonSerialized] internal int m_DevicesCount = -1;
-        [NonSerialized] internal InputDevice[] m_DevicesArray;
+        [NonSerialized] internal DeviceArray m_Devices;
 
         [NonSerialized] internal InlinedArray<Action<InputAction.CallbackContext>> m_ActionCallbacks;
 
         internal static int s_DeferBindingResolution;
 
+        internal struct DeviceArray
+        {
+            private bool m_HaveValue;
+            private int m_DeviceCount;
+            private InputDevice[] m_DeviceArray; // May have extra capacity; we won't let go once allocated.
+
+            public int IndexOf(InputDevice device)
+            {
+                if (m_DeviceCount > 0)
+                    return m_DeviceArray.IndexOfReference(device, m_DeviceCount);
+                return -1;
+            }
+
+            public bool Remove(InputDevice device)
+            {
+                var index = IndexOf(device);
+                if (index < 0)
+                    return false;
+                m_DeviceArray.EraseAtWithCapacity(ref m_DeviceCount, index);
+                return true;
+            }
+
+            public ReadOnlyArray<InputDevice>? Get()
+            {
+                if (!m_HaveValue)
+                    return null;
+                return new ReadOnlyArray<InputDevice>(m_DeviceArray, 0, m_DeviceCount);
+            }
+
+            public bool Set(ReadOnlyArray<InputDevice>? devices)
```

Comment:
Feels like this method is doing too much. Might be worth splitting this into another method called ContainsAll that can do the check for whether all devices exist in m_DeviceArray, and not having Set return a bool. Also, this is currently more of a Sequence Equals than a contains, so you only skip rebinding if the devices exist and are in the same order. That intentional?

Reply:
(fuck... buttons are my friends...)

Moved the check into a method.

>Also, this is currently more of a Sequence Equals than a contains, so you only skip rebinding if the devices exist and are in the same order. That intentional?

Sort of. Somewhat gray territory. Technically, it should be a set rather than a sequence. So order should be ignored.

However, when it comes to binding resolution, the devices will be tried in the order they occur. So in the end, order does end up having a subtle influence (which in practice rarely ever matters).

So from that POV, not sure. Ended up leaving it alone.

Reply:
I'm a fan of command/query separation, and I'd like if the Set method didn't return anything to fit with that. It's not at all clear what that bool will represent without reading the code...but this is something for another day. PR looks good.